django/cantusdb_project/main_app/management/commands/update_cached_concordances.py

```diff
@@ -0,0 +1,100 @@
+import ujson
+import os
+from sys import stdout
+from datetime import datetime
+from collections import defaultdict
+from django.db.models.query import QuerySet
+from django.core.management.base import BaseCommand
+from main_app.models import Chant
+
+
+class Command(BaseCommand):
+    def handle(self, *args, **kwargs) -> None:
+        CACHE_DIR: str = "api_cache"
+        FILEPATH: str = "api_cache/concordances.json"
+        start_time: str = datetime.now().isoformat()
+        stdout.write(f"Running update_cached_concordances at {start_time}.\n")
+        concordances: dict = get_concordances()
+        write_time: str = datetime.now().isoformat()
+        metadata: dict = {
+            "last_updated": write_time,
+        }
+        data_and_metadata: dict = {
+            "data": concordances,
+            "metadata": metadata,
+        }
+        stdout.write(f"Attempting to make directory at {CACHE_DIR} to hold cache: ")
+        try:
+            os.mkdir(CACHE_DIR)
+            stdout.write(f"successfully created directory at {CACHE_DIR}.\n")
+        except FileExistsError:
+            stdout.write(f"directory at {CACHE_DIR} already exists.\n")
+        stdout.write(f"Writing concordances to {FILEPATH} at {write_time}.\n")
+        with open(FILEPATH, "w") as json_file:
+            ujson.dump(data_and_metadata, json_file)
+        end_time = datetime.now().isoformat()
+        stdout.write(
+            f"Concordances successfully written to {FILEPATH} at {end_time}.\n\n"
+        )
+
+
+def get_concordances() -> dict:
+    DOMAIN: str = "https://cantusdatabase.org"
+
+    stdout.write("Querying database for published chants\n")
+    published_chants: QuerySet[Chant] = Chant.objects.filter(source__published=True)
+    values: QuerySet[dict] = published_chants.select_related(
```

Comment:
Not critical, but I'm confused by this type annotation... a `QuerySet` of dicts?

Reply:
hm... once we've used `.values()`, the items within the queryset are `dicts` instead of `Chant` objects. I almost considered casting `values` here to a list, so I could annotate it `list[dict]`, but that seemed unnecessary.

I'm gathering you think it would be clearer to annotate it as `QuerySet[Chant]` again - is this a change you'd like me to make?

Reply:
Going to merge this so that I can clean up something I noticed in `requirements.txt` without creating merge conflicts. But I'd be happy to make this change in another small PR if you think it would improve the function.

Reply:
If your type checker fine with QuerySet[dict] that must be it...it just seems odd. Iterating through the QuerySet gives you dicts, but I think it would still be a QuerySet of a particular model

Reply:
> Returns a QuerySet that returns dictionaries, rather than model instances, when used as an iterable.

https://docs.djangoproject.com/en/5.0/ref/models/querysets/#values

The type annotation is correct, I think.